New mutable hash map with Long keys: partially solves SI-5263 and is relevant to SI-6825.

Ichoran opened the pull request:
This implements a new hash map, mutable.LongMap, based on an open addressing scheme that provides
dramatically faster (mostly due to specialization) get and contains
operations than either the standard Java HashMap or any of the existing
Scala hash maps.  It doesn't work well above 500,000,000 elements as
the arrays cannot scale past 2^30 elements.  Maps are not faster in
general due to the lack of specialization of Function1[Long, V].

Ichoran committed:
New mutable hash map with Long keys: partially solves SI-5263 and is relevant
to SI-6825.

The hash map is based on an open addressing scheme that provides
dramatically faster (mostly due to specialization) get and contains
operations than either the standard Java HashMap or any of the existing
Scala hash maps.  It doesn't work well above 500,000,000 elements as
the arrays cannot scale past 2^30 elements.  Maps are not faster in
general due to the lack of specialization of Function1[Long, V].

Revisions made sure that all return types in the public API are specified.
Also switched to a leading-zeros method of calculating the initial mask (to
save a few ns), and used an occasionally-more-efficient version of
seekEntryOrOpen.  Also edited out specific performance numbers and moved
constants to companion.

Ichoran commented:
Some subset of @JamesIry and @gkossakowski and @adriaanm might want to look this over.

Ichoran committed:
New AnyRefMap fixes SI-5263 to the extent practical.

An open addressing hash map based on a similar scheme to mutable.LongMap.
It delivers performance equivalent to Java's HashMap for pretty much all
AnyRefs, plus it works nicely with Scala's collections.

Revisions made sure that all return types in the public API are specified.
Also switched to a leading-zeros method of calculating the initial mask (to
save a few ns).  Also edited out java.util comparison numbers and moved
constants to companion.

adriaanm commented:
Please rebase this on current master. In your `issue/5263` branch, do `git pull --rebase https://github.com/scala/scala.git master` (or abbreviate using your name for the scala remote)

adriaanm commented:
(and then do a `git push -f`)

Ichoran commented:
@adriaanm - Okay, rebased.

Ichoran commented:
Is this failure related to anything in my pull request?

  https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/3575/console

I can't quite decipher the log output.

  --Rex

P.S. I marked LongMap and AnyRefMap as serializable, but forgot to actually
implement any serialization.

On Mon, Nov 11, 2013 at 12:07 PM, Adriaan Moors notifications@github.comwrote:

> (and then do a git push -f)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3119#issuecomment-28234007
> .

adriaanm commented:
You and me both :-) It's not related to your work -- I'm working on diagnosing/fixing right now, as tracked in the comments at #3114

adriaanm commented:
(https://github.com/scala/scala/pull/3114#issuecomment-28244039)

adriaanm commented:
Please rebase on master again -- should work for real now. Apologies for the noise.

adriaanm commented:
@JamesIry, could you take this one?

JamesIry commented:
Yup

JamesIry commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -0,0 +1,456 @@
+  private[this] var myHashes: Array[Int] = null
+  private[this] var myKeys: Array[Object] = null
+  private[this] var myValues: Array[Object] = null
+    
```
Our convention is to use "size" rather than "mySize" and use "_size" if "size" conflicts with the name of a method. Same with "myKeys", "myInitialize", etc

Ichoran replied:
Okay.  I figured since Martin thought it was interesting and would "try it out" that it might be okay, but I can prefix with _ instead.  I just find it visually distracting since _ means other important things.

JamesIry commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -0,0 +1,456 @@
+ *
+ */
+final class AnyRefMap[K <: Object, V] private[collection] (defaultEntry: K => V, initialBufferSize: Int, initBlank: Boolean)
+extends AbstractMap[K, V]
```
K <: AnyRef would be more in keeping with Scala. That's a general comment for all uses of Object.

Ichoran replied:
Indeed.  Object is easier to type and makes me think of the right Java-y things.  Now that I'm done with building the class I'm happy to s/Object/AnyRef.

retronym commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -0,0 +1,456 @@
+ *  rapidly as 2^30 is approached.
+ *
+ */
+final class AnyRefMap[K <: Object, V] private[collection] (defaultEntry: K => V, initialBufferSize: Int, initBlank: Boolean)
```
Yay for `final`!

retronym commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -0,0 +1,456 @@
+    this
+  }
+  
+  def iterator = new Iterator[(K, V)] {
```
This should be type annotated, to avoid wondering whether we'll end up with a refinement type. As a rule, type annotate all public members.

retronym commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -0,0 +1,456 @@
+   *  Unlike `mapValues`, this method generates a new
+   *  collection immediately.
+   */
+  def mapValuesNow[V1](f: V => V1): AnyRefMap[K, V1] = {
```
Not sure about the name; might be the first use of `now` in the collections.

Maybe `mapValuesForce`? Anyone else have suggestions here?

Ichoran replied:
Suggestions would be most welcome.  There is no standard way to denote lazy vs. strict operations, and the library is not written such that you even have .force available every time you have a lazy operation (e.g. with normal `mapValues`).

retronym commented on src/library/scala/collection/mutable/LongMap.scala:
```diff
@@ -0,0 +1,553 @@
+}
+
+object LongMap {
+  private[collection] val exceptionDefault = (k: Long) => throw new NoSuchElementException(k.toString)
```
Could this (and its copy, `AnyRefMap.exceptionDefault`) be simply `private`? Or can we share one instance between both maps?

Ichoran replied:
I think plain `private`--or not private at all!--is the way to go.  If `AnyRef` specialization is ever recovered, you'll get much better performance off of defaults if the primitive and non-primitive maps don't share an implementation.  In favor of private: nobody else needs it.  In favor of public: you don't have to rewrite exceptionDefault if you want to have a menu of defaults to choose from.

retronym replied:
When in doubt, err on the side of privacy...

retronym commented on src/library/scala/collection/mutable/LongMap.scala:
```diff
@@ -0,0 +1,553 @@
+object LongMap {
+  private[collection] val exceptionDefault = (k: Long) => throw new NoSuchElementException(k.toString)
+  
+  implicit def canBuildFrom[V, U] = new CanBuildFrom[LongMap[V], (Long, U), LongMap[U]] {
```
Implicits should be type annotated; this might be discarded otherwise in implicit searches earlier in this compilation unit.

retronym replied:
Although, looking around, I see you this isn't done everywhere. Boy scout badged on offer to fix this in `GenSet`, `GenIterable`, `SeqView`, `IntMap`, and no doubt elsewhere.

retronym commented on src/library/scala/collection/mutable/LongMap.scala:
```diff
@@ -0,0 +1,553 @@
+    val sz = if (elems.hasDefiniteSize) elems.size else 4
+    val lm = new LongMap[V](sz * 2)
+    elems.foreach{ case (k,v) => lm(k) = v }
+    if (lm.size < (sz>>3)) lm.repack()
```
I've got a preference for whitespace around all operators. I don't know if we have an official policy on that way, but it seems to be the most amenable to something that we could get through an automatic formatter, which I'd like to introduce.

retronym commented on src/library/scala/collection/mutable/LongMap.scala:
```diff
@@ -0,0 +1,553 @@
+  /** Creates a new `LongMap` from keys and values. 
+   *  Equivalent to but more efficient than `LongMap((keys zip values): _*)`.
+   */
+  def fromZip[V](keys: Iterable[Long], values: Iterable[V]) = {
```
A few more candidates for type annotation

Ichoran commented:
@JamesIry @retronym @adriaanm - Addressed all comments, I believe.  (Except the whitespaced operators; I prefer that whitespace visually indicate precedence, so 5*x + 2 not 5 \* x+2.  I would love a tool that would yell at me when my whitespace contradicts operator precedence.  Anyway, since Jason mentioned that a formatter would be the way to do it, I didn't feel compelled to find by hand spots where I omitted whitespace.)

adriaanm commented:
We prefer JUnit test for library functionality. Here's a recent example: https://github.com/scala/scala/pull/3138/files#diff-5

Use ant.junit to run the test suite.  (Edit: I realize I'm late to the party. This can go on a TODO list somewhere.)

adriaanm commented:
All, I'm tagging M7 at 10:30 SF time tomorrow, Monday. It's okay if this isn't ready to be merged by then, it can go into M8 (though it would of course be nice to have!)

Ichoran commented:
Is it required to switch tests to JUnit given that the other collections tests, whose style I copied, were not?  If JUnit is required then I might not be able to finish tonight.  (Not sure about LGTMs, but I'll have all other input addressed by then.)

adriaanm commented:
I won't insist on JUnit, though I would like to shift the tide towards it, especially for the library.

JamesIry commented:
LGTM but I'd like to see a follow up commit moving to JUnit

adriaanm commented:
Thanks guys!
